tests/rptest/tests/alter_topic_configuration_test.py

```diff
@@ -323,3 +353,137 @@ def check():
+
+        install_opts = InstallOptions(version=RedpandaVersionTriple(
+            (24, 1, 1)),
+                                      num_to_upgrade=2)
+        self.start_redpanda(
+            num_nodes=num_nodes,
+            si_settings=SISettings(test_context=self.test_context),
+            install_opts=install_opts)
+
+        redpanda_versions = {
+            i: self.redpanda.get_version_int_tuple(node)
+            for (i, node) in enumerate(self.redpanda.nodes)
+        }
+
+        rpk = RpkTool(self.redpanda)
+        # KCL is used to direct AlterConfig and DescribeConfigs requests to specific brokers.
+        kcl = KCL(self.redpanda)
+        topic = self.topics[0].name
+        props = {
+            'redpanda.remote.read': 'false',
+            'redpanda.remote.write': 'false'
+        }
+
+        rpk.create_topic(topic, partitions=1, replicas=3, config=props)
+
+        # Make sure that the lower versioned node is the partition leader.
+        leader_node_index = min(redpanda_versions, key=redpanda_versions.get)
+        leader_node_id = leader_node_index + 1
+        self.redpanda._admin.partition_transfer_leadership(
+            'kafka', topic, 0, leader_node_id)
+        wait_until(lambda: self.redpanda._admin.get_partition_leader(
+            namespace='kafka', topic=topic, partition=0) == leader_node_id,
+                   timeout_sec=10,
+                   backoff_sec=1,
+                   err_msg="Partition leadership did not stabilize")
+
+        # Make sure that the higher versioned node is the controller.
+        controller_node_index = max(redpanda_versions,
+                                    key=redpanda_versions.get)
+        controller_node_id = controller_node_index + 1
+        self.redpanda._admin.partition_transfer_leadership(
+            'redpanda', 'controller', 0, controller_node_id)
+        wait_until(lambda: self.redpanda._admin.get_partition_leader(
+            namespace="redpanda", topic="controller", partition=0) ==
+                   controller_node_id,
+                   timeout_sec=10,
+                   backoff_sec=1,
+                   err_msg="Controller leadership did not stabilize")
+
+        # Sanity check defaults
+        desc = rpk.describe_topic_configs(topic)
+        assert desc['redpanda.remote.read'][0] == 'false'
+        assert desc['redpanda.remote.write'][0] == 'false'
+
+        third_node_index = [
+            i for i in redpanda_versions.keys()
+            if i not in [controller_node_index, leader_node_index]
+        ][0]
+
+        controller_node = self.redpanda.nodes[controller_node_index]
+        leader_node = self.redpanda.nodes[leader_node_index]
+        third_node = self.redpanda.nodes[third_node_index]
+
+        # The deprecated code path is buggy, so we
+        # can only assert topic properties on upgraded nodes.
+        upgraded_nodes = [controller_node, third_node]
+
+        def assert_remote_read_and_write_on_upgraded_nodes(
```

Comment:
Added assert on all nodes, with a [stipulation on the final, expectedly buggy `AlterConfig` request to the un-upgraded node](https://github.com/redpanda-data/redpanda/compare/9d34ec810aaba69962f6adf12fac0f1d4d796e77..32eb8372ca2c7b376490dc5e40109de347a0ef24#diff-0879970555c08129c60c644a81e0f1a62799e61d8e9f026089502075e5db4235R508-R513).

`v24.1` is [lacking some updates made to `AlterConfig`](https://github.com/redpanda-data/redpanda/pull/21291/commits/532f699816433b3e653a4b38f7ce123003b1bedf), and we cannot `assert` on a consistent state.

When `v24.3` is officially cut, we can bump the un-upgraded version used in this test to `v24.2` and remove the work-around added here.

Reply:
The above comment is incorrect- for some reason but not the reason stated above, we see an inconsistent state on the un-upgraded node, which is of version `v24.1.1`.

I don't think the inconsistency in topic properties between nodes is introduced by any changes made in this PR. 

There are changes to the `controller_backend` and [`shard_placement_table`](https://github.com/redpanda-data/redpanda/blame/338e9bd03afdb32f4d2cc3ab9de38ff78c2387c1/src/v/cluster/shard_placement_table.cc) that were not backported to `v24.1`, so whatever bug is present in that version is hard to track down.

I think that the test should be reverted to only asserting on the upgraded nodes. When `v24.3` is cut, we can bump the un-upgraded version to `v24.2` and see if we can add the assert on the lower version node as well, assuming the unknown bug isn't present in that version.

Reply:
I've had a deeper look this morning and it seems that the reason for the inconsistency is that on new nodes, the update encoded into the `shadow_indexing` field is inconsistent with the update encoded into the new fields. We could try making them "more often consistent" (see my local testing here: https://github.com/redpanda-data/redpanda/commit/1ee94d607d6c72c226e44d584b4fc17de5aff63c), but we cannot make them always consistent. In fact, the reason why we cannot make them fully consistent is the same reason why we are introducing the two new fields in the first place, which is that `redpanda.remote.read=false` and `redpanda.remote.write=true` can't be expressed with the `shadow_indexing_mode` enum.

So I believe the best way forward would be to move the feature switch up to the kafka handlers and implement it with a feature gate that gets enabled once the cluster is fully upgraded. It simplifies things because it means that we don't have to deal with forward compatibility in a mixed-mode cluster, and we only have to solve for backward compatibility by keeping the code that handles old-style commands in place inside `topics_table`.

Using a feature gate also means that we wouldn't be able to backport this change, which seems acceptable if not even desirable. I certainly think that ensuring that the nodes have a consistent view of the topic configs is more important here than being able to backport this change.

This also simplifies testing slightly since there is a clear cutoff point between the old behaviour and the new behaviour. A mixed-mode cluster would work the same as an old cluster, and we only need to verify that an upgraded cluster can still process old-style commands.

Reply:
I have added the new feature flag `shadow_indexing_split_topic_property_update`, which is used at the `kafka` layer in order to determine which code-path should be followed for the `shadow_indexing` topic property update- the old, bug prone code path, or the updated one.